### PR TITLE
Refactor errors: cli_wallet #804

### PR DESCRIPTION
**Pull request by s-medvedev**

### CODE REVIEW

File: libraries/wallet/wallet.cpp [hunk 1]
```diff
@@ -1755,41 +1802,37 @@ fc::ecc::private_key wallet_api::derive_private_key(const std::string& prefix_st
...
         {
-            FC_ASSERT( !is_locked() );
+            WALLET_CHECK_UNLOCKED();
 
-            auto accounts = my->_remote_database_api->get_accounts( { account_name } );
-            FC_ASSERT( accounts.size() == 1, "Account does not exist" );
-            FC_ASSERT( account_name == accounts[0].name, "Account name doesn't match?" );
-            FC_ASSERT( threshold != 0, "Authority is implicitly satisfied" );
+            auto account = get_account(account_name);
+            GOLOS_CHECK_PARAM(threshold, GOLOS_CHECK_VALUE(threshold != 0, "Authority is implicitly satisfied"));
```
Comment (zxcat) [inline libraries/wallet/wallet.cpp]: Moving this check before `get_account` will remove unnecessary remote request in case of bad `threshold`

File: libraries/wallet/wallet.cpp [hunk 2]
```diff
@@ -1860,11 +1900,7 @@ fc::ecc::private_key wallet_api::derive_private_key(const std::string& prefix_st
         }
 
         annotated_signed_transaction wallet_api::delegate_vesting_shares(string delegator, string delegatee, asset vesting_shares, bool broadcast) {
-            FC_ASSERT(!is_locked());
-            auto accounts = my->_remote_database_api->get_accounts({delegator, delegatee});
-            FC_ASSERT(accounts.size() == 2, "One or more of the accounts specified do not exist.");
-            FC_ASSERT(delegator == accounts[0].name, "Delegator account is not right?");
-            FC_ASSERT(delegatee == accounts[1].name, "Delegatee account is not right?");
```
Comment (zxcat) [inline libraries/wallet/wallet.cpp]: Why did you remove this checks?

File: libraries/wallet/wallet.cpp [hunk 3]
```diff
@@ -2042,30 +2078,45 @@ fc::ecc::private_key wallet_api::derive_private_key(const std::string& prefix_st
             for( auto& key_weight_pair : account.owner.key_auths )
             {
                 for( auto& key : keys )
-                    FC_ASSERT( key_weight_pair.first != key, "Detected private owner key in memo field. Cancelling transaction." );
+                    GOLOS_CHECK_LOGIC(key_weight_pair.first != key, 
+                            logic_errors::detected_private_key_in_memo,
+                            "Detected ${type} private key in memo field",
+                            ("type","owner"));
```
Comment (zxcat) [inline libraries/wallet/wallet.cpp]: Why use constant string as substitution? Result will be always the same

File: libraries/wallet/wallet.cpp [hunk 4]
```diff
@@ -2042,30 +2078,45 @@ fc::ecc::private_key wallet_api::derive_private_key(const std::string& prefix_st
...
             }
 
             for( auto& key_weight_pair : account.active.key_auths )
             {
                 for( auto& key : keys )
-                    FC_ASSERT( key_weight_pair.first != key, "Detected private active key in memo field. Cancelling transaction." );
+                    GOLOS_CHECK_LOGIC(key_weight_pair.first != key, 
+                            logic_errors::detected_private_key_in_memo,
+                            "Detected ${type} private key in memo field",
+                            ("type","active"));
```
Comment (zxcat) [inline libraries/wallet/wallet.cpp]: Another string generation with already known result

File: libraries/wallet/wallet.cpp [hunk 5]
```diff
@@ -2042,30 +2078,45 @@ fc::ecc::private_key wallet_api::derive_private_key(const std::string& prefix_st
...
             }
 
             for( auto& key_weight_pair : account.posting.key_auths )
             {
                 for( auto& key : keys )
-                    FC_ASSERT( key_weight_pair.first != key, "Detected private posting key in memo field. Cancelling transaction." );
+                    GOLOS_CHECK_LOGIC(key_weight_pair.first != key, 
+                            logic_errors::detected_private_key_in_memo,
+                            "Detected ${type} private key in memo field",
+                            ("type","posting"));
```
Comment (zxcat) [inline libraries/wallet/wallet.cpp]: Another string generation with already known result

File: libraries/wallet/wallet.cpp [hunk 6]
```diff
@@ -2042,30 +2078,45 @@ fc::ecc::private_key wallet_api::derive_private_key(const std::string& prefix_st
...
             }
 
             const auto& memo_key = account.memo_key;
             for( auto& key : keys )
-                FC_ASSERT( memo_key != key, "Detected private memo key in memo field. Cancelling transaction." );
+                GOLOS_CHECK_LOGIC(memo_key != key, 
+                        logic_errors::detected_private_key_in_memo,
+                        "Detected ${type} private key in memo field",
+                        ("type","memo"));
```
Comment (zxcat) [inline libraries/wallet/wallet.cpp]: Another string generation with already known result

File: libraries/wallet/wallet.cpp [hunk 7]
```diff
@@ -2042,30 +2078,45 @@ fc::ecc::private_key wallet_api::derive_private_key(const std::string& prefix_st
...
 
             // Check against imported keys
             for( auto& key_pair : my->_keys )
             {
                 for( auto& key : keys )
-                    FC_ASSERT( key != key_pair.first, "Detected imported private key in memo field. Cancelling trasanction." );
+                    GOLOS_CHECK_LOGIC(key_pair.first != key, 
+                            logic_errors::detected_private_key_in_memo,
+                            "Detected ${type} private key in memo field",
+                            ("type","imported"));
```
Comment (zxcat) [inline libraries/wallet/wallet.cpp]: Another string generation with already known result

File: libraries/wallet/wallet.cpp [hunk 8]
```diff
@@ -2435,7 +2486,6 @@ fc::ecc::private_key wallet_api::derive_private_key(const std::string& prefix_st
         }
 
         market_history::order_book wallet_api::get_order_book( uint32_t limit ) {
-            FC_ASSERT( limit <= 1000 );
```
Comment (zxcat) [inline libraries/wallet/wallet.cpp]: Why?

Comment (s-medvedev) [reply libraries/wallet/wallet.cpp]: This assert duplicate code on server side (there this parameter is limited with 500).

File: libraries/wallet/wallet.cpp (hunk 3 above)
Comment (zxcat) [reply libraries/wallet/wallet.cpp]: I mean why use
`"Detected ${type} private key in memo field", ("type","owner")`
vs simple
`"Detected owner private key in memo field"`

File: libraries/wallet/wallet.cpp (hunk 2 above)
Comment (s-medvedev) [reply libraries/wallet/wallet.cpp]: In the other places this check is necessary, because the account data is used for the correct creation of the request.

Comment (zxcat) [reply libraries/wallet/wallet.cpp]: So as I understood right, all other checks don't removed (`account = get_account()`) only because `account` variable used later in code. Ok.

Comment (zxcat) [reply libraries/wallet/wallet.cpp]: > In the other places this check is necessary, because the account data is used for the correct creation of the request.

just wrote the same : )

File: libraries/wallet/wallet.cpp (hunk 3 above)
Comment (zxcat) [reply libraries/wallet/wallet.cpp]: > Passing this parameter this way not only performs substitution, but also adds this parameter with a value to the list of error properties

Ok, I see. If so, maybe put it in lambda or function to remove slightly reduce duplication and make this decision clearer.

File: libraries/wallet/wallet.cpp (hunk 1 above)
Comment (s-medvedev) [reply libraries/wallet/wallet.cpp]: Fixed.
